[JSC] Spew strict-eq Baseline JIT code with constant strings

Constellation opened the pull request:
#### a36c2519d4eed1a1764f6f050ee77e30dfa58b8c
<pre>
[JSC] Spew strict-eq Baseline JIT code with constant strings
<a href="https://bugs.webkit.org/show_bug.cgi?id=269106">https://bugs.webkit.org/show_bug.cgi?id=269106</a>
<a href="https://rdar.apple.com/122674588">rdar://122674588</a>

Reviewed by Alexey Shvayka.

Let&apos;s leverage the fact that there are many `&quot;string&quot; === x` comparisons.
In that case, we can emit very specific optimized code even in Baseline JIT easily.
This patch adds `StringIdent === x` case optimizations in Baseline JIT.

Furthermore, we found that

```
    switch (expr) {
    case &quot;string1&quot;:
        ...
    case &quot;string2&quot;:
        ...
    case variable:
        ...
    }
```

case is emitting very inefficient bytecode, which does not use constant register directly with `jstricteq`.
As a result, my new optimization does not kick in with this. This patch also fixes BytecodeGenerator to make
this new optimization work well by emitting `jstricteq constant, x`.

* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::compileOpStrictEq):
(JSC::JIT::compileOpStrictEqJump):

Canonical link: <a href="https://commits.webkit.org/274418@main">https://commits.webkit.org/274418@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a055d658cfae8e16f0d147a1d5a5a5f68ee56d4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39048 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17981 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41401 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41582 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20864 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15329 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32679 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39622 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15150 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33848 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13157 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13112 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34783 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42859 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/32501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35446 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35111 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38944 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/38674 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13860 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11430 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37172 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15466 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/45680 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15128 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9309 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5100 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14952 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shvaikalesh reviewed (approved):
r=me yet last sentence of ChangeLog may needs a tweak?

Constellation committed:
[JSC] Spew strict-eq Baseline JIT code with constant strings
https://bugs.webkit.org/show_bug.cgi?id=269106
rdar://122674588

Reviewed by Alexey Shvayka.

Let's leverage the fact that there are many `"string" === x` comparisons.
In that case, we can emit very specific optimized code even in Baseline JIT easily.
This patch adds `StringIdent === x` case optimizations in Baseline JIT.

Furthermore, we found that

```
    switch (expr) {
    case "string1":
        ...
    case "string2":
        ...
    case variable:
        ...
    }
```

case is emitting very inefficient bytecode, which does not use constant register directly with `jstricteq`.
As a result, my new optimization does not kick in with this. This patch also fixes BytecodeGenerator to make
this new optimization work well by emitting `jstricteq constant, x`.

* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::compileOpStrictEq):
(JSC::JIT::compileOpStrictEqJump):

Canonical link: https://commits.webkit.org/274418@main
